File: app/src/main/java/com/hellmund/primetime/history/HistoryItemTouchHelper.java
package com.hellmund.primetime.history;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.View;

class HistoryItemTouchHelper extends ItemTouchHelper.Callback {

    private HistoryAdapter mAdapter;
    private Drawable mBackground;

    HistoryItemTouchHelper(Context context, HistoryAdapter adapter) {
        mAdapter = adapter;
        mBackground = ContextCompat.getDrawable(context, android.R.color.transparent);
    }

    @Override
    public int getMovementFlags(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
        return makeFlag(ItemTouchHelper.ACTION_STATE_SWIPE, ItemTouchHelper.END)
                | makeFlag(ItemTouchHelper.ACTION_STATE_SWIPE, ItemTouchHelper.START);
    }

    @Override
    public boolean isItemViewSwipeEnabled() {
        return true;
    }

    @Override
    public boolean onMove(RecyclerView recyclerView,
                          RecyclerView.ViewHolder viewHolder, RecyclerView.ViewHolder target) {
        return false;
    }

    @Override
    public void onSwiped(RecyclerView.ViewHolder holder, int direction) {
        final int position = holder.getAdapterPosition();
        mAdapter.removeItem(position);
    }

    @Override
    public void onChildDraw(Canvas c, RecyclerView recyclerView,
                            RecyclerView.ViewHolder holder, float dX, float dY,
                            int actionState, boolean isCurrentlyActive) {
        if (holder.getAdapterPosition() == -1) {
            return;
        }

        final float width = holder.itemView.getWidth();
        final float percentage = 1 - (Math.abs(dX) / width);
        holder.itemView.setAlpha(percentage);

        mBackground.setBounds(getBackgroundRect((int) dX, holder.itemView));
        mBackground.draw(c);

        super.onChildDraw(c, recyclerView, holder, dX, dY, actionState, isCurrentlyActive);
    }

    private Rect getBackgroundRect(int dX, View view) {
        final int left = view.getLeft();
        final int top = view.getTop();
        final int right = view.getLeft() + dX;
        final int bottom = view.getBottom();
        return new Rect(left, top, right, bottom);
    }

}

File: app/src/main/java/com/hellmund/primetime/ui/history/HistoryItemTouchHelper.kt
package com.hellmund.primetime.ui.history

import android.content.Context
import android.graphics.Canvas
import android.graphics.Rect
import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.support.v7.widget.helper.ItemTouchHelper.*
import android.view.View
import java.lang.Math.abs

internal class HistoryItemTouchHelper(
        context: Context,
        private val adapter: HistoryAdapter
) : ItemTouchHelper.Callback() {

    private val background = ContextCompat.getDrawable(context, android.R.color.transparent)

    override fun getMovementFlags(
            recyclerView: RecyclerView,
            viewHolder: RecyclerView.ViewHolder
    ): Int = makeFlag(ACTION_STATE_SWIPE, END) or makeFlag(ACTION_STATE_SWIPE, START)

    override fun isItemViewSwipeEnabled(): Boolean {
        return true
    }

    override fun onMove(
            recyclerView: RecyclerView,
            viewHolder: RecyclerView.ViewHolder,
            target: RecyclerView.ViewHolder
    ): Boolean = false

    override fun onSwiped(holder: RecyclerView.ViewHolder, direction: Int) {
        val position = holder.adapterPosition
        adapter.removeItem(position)
    }

    override fun onChildDraw(c: Canvas, recyclerView: RecyclerView,
                             holder: RecyclerView.ViewHolder, dX: Float, dY: Float,
                             actionState: Int, isCurrentlyActive: Boolean) {
        if (holder.adapterPosition == -1) {
            return
        }

        val width = holder.itemView.width.toFloat()
        val percentage = 1 - abs(dX) / width
        holder.itemView.alpha = percentage

        background?.bounds = holder.itemView.rect.add(right = dX.toInt())
        background?.draw(c)

        super.onChildDraw(c, recyclerView, holder, dX, dY, actionState, isCurrentlyActive)
    }

    private val View.rect: Rect
        get() = Rect(left, top, right, bottom)

    private fun Rect.add(
            left: Int = 0,
            top: Int = 0,
            right: Int = 0,
            bottom: Int = 0
    ) = Rect(this.left + left, this.top + top, this.right + right, this.bottom + bottom)

}
